### PR TITLE
BASE: Add --dump-all-detection-entries command

### PR DESCRIPTION
**Pull request by PhoenixFlame101**
`--dump-all-detection-entries` is a command that prints formatted detection entries to STDOUT.

Currently, it works for engines using `AdvancedDetector`, which includes all but 3 engines - Glk, Sky, and SCUMM.

### CODE REVIEW

**Comment by i30817** [comment]
This will be useful to add thumbnails to libretro-database, especially combined with my project to fuzz thumbnail names (which recently got a way to use a local server) thanks. 


It was a pain to follow the detection.h files to 'deduce' the possible versions, and i'm sure i missed some, pity about scumm engine not being included

File: base/commandLine.cpp [hunk 1]
```diff
@@ -69,190 +70,190 @@ static const char USAGE_STRING[] =
...
-	"  --game=ID                In combination with --add or --detect only adds or attempts to\n"
-	"                           detect the game with id ID.\n"
-	"  --engine=ID              In combination with --list-games, --list-all-games, --list-targets, or --stats only\n"
-	"                           lists games for this engine. Multiple engines can be listed separated by a coma.\n"
-	"  --auto-detect            Display a list of games from current or specified directory\n"
-	"                           and start the first one. Use --path=PATH to specify a directory.\n"
-	"  --recursive              In combination with --add or --detect recurse down all subdirectories\n"
-	"  --no-exit                In combination with commands that exit after running, like --add or --list-engines,\n"
-	"                           open the launcher instead of exiting\n"
+	"  -v, --version            	Display ScummVM version information and exit\n"
```
Comment (sev-) [inline base/commandLine.cpp:73]: Please revert this while the indentation change and make the new line just protrude further, e.g. like `--output-channels=CHANNELS`

File: base/commandLine.cpp [hunk 2]
```diff
@@ -31,6 +31,7 @@
 #include "base/version.h"
 
 #include "common/config-manager.h"
+#include "common/file.h"
```
Comment (sev-) [inline base/commandLine.cpp:34]: This include is a leftover from the previous approach and must be removed

File: base/commandLine.cpp [hunk 3]
```diff
@@ -1419,6 +1423,21 @@ static void listAudioDevices() {
 	}
 }
 
+/** Dump MD5s from detection entries into STDOUT */
+static void dumpAllDetectionEntries() {
+	const PluginList &plugins = EngineMan.getPlugins();
+
+	debug("scummvm (");
```
Comment (sev-) [inline base/commandLine.cpp:1430]: Please use `printf()` here

File: engines/advancedDetector.cpp [hunk 4]
```diff
@@ -626,6 +651,54 @@ static bool getFilePropertiesIntern(uint md5Bytes, const AdvancedMetaEngine::Fil
...
+}
+
+void AdvancedMetaEngineDetection::dumpDetectionEntries() const {
+	const byte *descPtr;
+
+	for (descPtr = _gameDescriptors; ((const ADGameDescription *)descPtr)->gameId != nullptr; descPtr += _descItemSize) {
+		auto g = ((const ADGameDescription *)descPtr);
+		const char *title = ((const PlainGameDescriptor *)_gameIds)->description;
+
+		debug("game (");
```
Comment (sev-) [inline engines/advancedDetector.cpp:678]: Please use `printf()` here, thus add `#define FORBIDDEN_SYMBOL_EXCEPTION_printf` at the beginning of the file

File: engines/advancedDetector.cpp [hunk 5]
```diff
@@ -626,6 +651,54 @@ static bool getFilePropertiesIntern(uint md5Bytes, const AdvancedMetaEngine::Fil
...
+		debug("\tplatform %s", escapeString(getPlatformCode(g->platform)).c_str());
+		debug("\tsourcefile %s", escapeString(getName()).c_str());
+
+		for (auto fileDesc = g->filesDescriptions; fileDesc->fileName; fileDesc++) {
+			Common::String fname = fileDesc->fileName;
+			int64 fsize = fileDesc->fileSize;
+			Common::String md5 = fileDesc->md5;
+			MD5Properties md5prop = gameFileToMD5Props(fileDesc, g->flags);
+			auto md5Prefix = Common::String(md5PropToGameFile(md5prop));
+			Common::String key = md5.c_str();
```
Comment (sev-) [inline engines/advancedDetector.cpp:692]: Why are you doing double conversion here? First, `md5` to `char *` and then `char *` back to string?

Just write `Common::String key = md5;`

File: engines/advancedDetector.cpp [hunk 6]
```diff
@@ -626,6 +651,54 @@ static bool getFilePropertiesIntern(uint md5Bytes, const AdvancedMetaEngine::Fil
...
+		debug("\tlanguage %s", escapeString(getLanguageLocale(g->language)).c_str());
+		debug("\tplatform %s", escapeString(getPlatformCode(g->platform)).c_str());
+		debug("\tsourcefile %s", escapeString(getName()).c_str());
+
+		for (auto fileDesc = g->filesDescriptions; fileDesc->fileName; fileDesc++) {
+			Common::String fname = fileDesc->fileName;
+			int64 fsize = fileDesc->fileSize;
+			Common::String md5 = fileDesc->md5;
+			MD5Properties md5prop = gameFileToMD5Props(fileDesc, g->flags);
+			auto md5Prefix = Common::String(md5PropToGameFile(md5prop));
```
Comment (sev-) [inline engines/advancedDetector.cpp:691]: Better write `Common::String md5Prefix = md5PropToGameFile(md5prop)`.

In general, this short piece of code is very messy: inconsistent usage of types, double conversion, etc.

File: engines/advancedDetector.cpp [hunk 7]
```diff
@@ -626,6 +651,54 @@ static bool getFilePropertiesIntern(uint md5Bytes, const AdvancedMetaEngine::Fil
...
+			Common::String key = md5.c_str();
+			if (md5Prefix != "" && md5.find(':') == Common::String::npos)
+				key = Common::String::format("%s:%s", md5Prefix.c_str(), md5.c_str());
```
Comment (sev-) [inline engines/advancedDetector.cpp:694]: `key = md5Prefix + ":" + md5;`

File: engines/advancedDetector.cpp [hunk 8]
```diff
@@ -626,6 +651,54 @@ static bool getFilePropertiesIntern(uint md5Bytes, const AdvancedMetaEngine::Fil
...
+
+void AdvancedMetaEngineDetection::dumpDetectionEntries() const {
+	const byte *descPtr;
+
+	for (descPtr = _gameDescriptors; ((const ADGameDescription *)descPtr)->gameId != nullptr; descPtr += _descItemSize) {
+		auto g = ((const ADGameDescription *)descPtr);
+		const char *title = ((const PlainGameDescriptor *)_gameIds)->description;
+
+		debug("game (");
+		debug("\tname %s", escapeString(g->gameId).c_str());
```
Comment (sev-) [inline engines/advancedDetector.cpp:679]: Aren't you supposed to put these escaped strings into quotes?

**Review by sev- — CHANGES_REQUESTED** [review]
Did my review. A lot of notes

File: engines/advancedDetector.cpp (hunk 8 above)
Comment (PhoenixFlame101) [reply engines/advancedDetector.cpp]: The quotes are so that we can account for strings with spaces in them. For metadata, the parser only sees the first whitespace, so this is not necessary.

File: base/commandLine.cpp [hunk 9]
```diff
@@ -24,11 +24,11 @@
 
 #define FORBIDDEN_SYMBOL_EXCEPTION_exit
 
-#include "engines/advancedDetector.h"
-#include "engines/metaengine.h"
 #include "base/commandLine.h"
 #include "base/plugins.h"
 #include "base/version.h"
+#include "engines/advancedDetector.h"
```
Comment (sev-) [inline base/commandLine.cpp:30]: What is the purpose of this reshuffling?

File: base/commandLine.cpp [hunk 10]
```diff
@@ -62,8 +62,7 @@ static const char USAGE_STRING[] =
 	"%s: %s\n"
 	"Usage: %s [OPTIONS]... [GAME]\n"
 	"\n"
-	"Try '%s --help' for more options.\n"
-;
```
Comment (sev-) [inline base/commandLine.cpp:66]: Why this was removed?

File: base/commandLine.cpp [hunk 11]
```diff
@@ -172,15 +172,15 @@ static const char HELP_STRING4[] =
 	"  --output-rate=RATE       Select output sample rate in Hz (e.g. 22050)\n"
 	"  --opl-driver=DRIVER      Select AdLib (OPL) emulator (db, mame"
 #ifndef DISABLE_NUKED_OPL
-																	 ", nuked"
```
Comment (sev-) [inline base/commandLine.cpp:175]: Please revert it properly

File: base/commandLine.cpp [hunk 12]
```diff
@@ -477,76 +475,82 @@ static Common::String createTemporaryTarget(const Common::String &engineId, cons
...
-		if (isLongCmd) { \
-			s += sizeof(longCmd) - 1; \
-			if (*s == '=') \
-				s++; \
-		} \
-		const char *option = s; \
-		if (*s == '\0' && !isLongCmd) { option = s2; i++; } \
-		if (!option || *option == '\0') option = defaultVal; \
-		if (option) settings[longCmd] = option;
+#define DO_OPTION_OPT(shortCmd, longCmd, defaultVal)                                                                                    \
```
Comment (sev-) [inline base/commandLine.cpp:478]: Please, do not do unnecessary whitespace changes

File: base/commandLine.cpp [hunk 13]
```diff
@@ -610,8 +614,8 @@ Common::String parseCommandLine(Common::StringMap &settings, int argc, const cha
 			END_COMMAND
 
 			DO_OPTION_OPT(0, "list-debugflags", "global")
-				ensureFirstCommand(command, "list-debugflags");
```
Comment (sev-) [inline base/commandLine.cpp:613]: What are you doing here?

**Review by sev- — CHANGES_REQUESTED** [review]
I put couple of comments, but `base/commandLine.cpp` file is basically trashed into pieces with this commit.

File: base/commandLine.cpp [hunk 14]
```diff
@@ -76,6 +75,7 @@ static const char HELP_STRING1[] =
 	"  -t, --list-targets       Display list of configured targets and exit\n"
 	"  --list-engines           Display list of supported engines and exit\n"
 	"  --list-all-engines       Display list of all detection engines and exit\n"
+	" --dump-all-detection-entries Create a DAT file containing MD5s from detection entries of all engines\n"
```
Comment (bluegr) [inline base/commandLine.cpp:78]:
```suggestion
	"  --dump-all-detection-entries Create a DAT file containing MD5s from detection entries of all engines\n"
```

File: engines/advancedDetector.cpp [hunk 15]
```diff
@@ -626,6 +653,54 @@ static bool getFilePropertiesIntern(uint md5Bytes, const AdvancedMetaEngine::Fil
...
+}
+
+void AdvancedMetaEngineDetection::dumpDetectionEntries() const {
+	const byte *descPtr;
+
+	for (descPtr = _gameDescriptors; ((const ADGameDescription *)descPtr)->gameId != nullptr; descPtr += _descItemSize) {
+		auto g = ((const ADGameDescription *)descPtr);
+		const char *title = ((const PlainGameDescriptor *)_gameIds)->description;
+
+		printf("game (");
```
Comment (sev-) [inline engines/advancedDetector.cpp:680]: Did you test the output after you changed debug() to print()?

**Comment by sev-** [comment]
Thanks!
